src/core/job.c

```diff
@@ -1498,80 +1473,42 @@ int job_get_before(Job *j, Job*** ret) {
                 return 0;
         }
 
-        HASHMAP_FOREACH_KEY(v, other, j->unit->dependencies[UNIT_AFTER], i) {
-                if (!other->job)
-                        continue;
-                if (job_compare(j, other->job, UNIT_AFTER) <= 0)
+        HASHMAP_FOREACH(other, j->manager->jobs, i) {
+                if (job_compare(j, other, _UNIT_DEPENDENCY_INVALID) <= 0)
                         continue;
 
                 if (!GREEDY_REALLOC(list, n_allocated, n+1))
                         return -ENOMEM;
-                list[n++] = other->job;
+                list[n++] = other;
         }
 
-        HASHMAP_FOREACH_KEY(v, other, j->unit->dependencies[UNIT_BEFORE], i) {
-                if (!other->job)
-                        continue;
-                if (job_compare(j, other->job, UNIT_BEFORE) <= 0)
-                        continue;
-
-                if (!GREEDY_REALLOC(list, n_allocated, n+1))
-                        return -ENOMEM;
-                list[n++] = other->job;
-        }
-
-        n = sort_job_list(list, n);
-
         *ret = TAKE_PTR(list);
 
         return (int) n;
 }
 
+/* Returns a list of all pending jobs that are waiting for this job to finish. */
 int job_get_after(Job *j, Job*** ret) {
         _cleanup_free_ Job** list = NULL;
         size_t n = 0, n_allocated = 0;
-        Unit *other = NULL;
-        void *v;
+        Job *other = NULL;
         Iterator i;
 
         assert(j);
         assert(ret);
 
-        /* Returns a list of all pending jobs that are waiting for this job to finish. */
-
-        HASHMAP_FOREACH_KEY(v, other, j->unit->dependencies[UNIT_BEFORE], i) {
-                if (!other->job)
-                        continue;
-
-                if (other->job->ignore_order)
+        HASHMAP_FOREACH(other, j->manager->jobs, i) {
+                if (other->ignore_order)
```

Comment:
I think this needs consideration. The reduction of code is nice, but it comes at a potential cost in computational complexity. Before, the search cost would be proportional to the size of the `dependencies[]` hashmaps, i.e. on the order of a few for most units, but now it is proportional to the size of the job list. That varies a lot, but some systems can have lots of units and jobs, e.g. ~1e5 e.g. for systems with a large number of devices. And now we have quadratic complexity over that ;(

Reply:
The foremost point -- job_get_{after,before} are function used in debugging DBus methods, not likely a hot path.

To the complexity: I assume hashmap lookup takes O(1), there are U units. Number of jobs J ∊ O(U), the number of potential dependencies for a single unit is in O(U), i.e. both loops (original, reworked) are equal. The original version needed the final sort to remove dupes, so overall we went down from O(U*log(U)) to O(U). What am I missing?

Reply:
If this function is called once for each job, then the complexity is O(J²). Before it was O(J×D), where D is the average number of dependencies. D is surprisingly low. E.g. on my machine there's just 3070 Before or After deps with 1034 loaded units. The sorting doesn't matter, because it's called on a very short list.

Your argument that this is just a debugging interface is correct. Nevertheless, I'd prefer not to do such a change. This is a public API, and we can't really know what people use it for.

Maybe you could rework this patch to avoid custom logic but still only iterate over  `j->unit->dependencies[UNIT_BEFORE]` and  `j->unit->dependencies[UNIT_AFTER]`? Otherwise, I'd prefer to drop this commit altogether.